tattle-made/Uli
# feat: slur crowdsource feature

**Describe the PR**
This PR creates a new feature where Slurs can be crowdsourced by Uli users. This PR resolves #316 and #289.

A crowdsourced slur will have the following metadata
1. label - slur word
2. label meaning - meaning of the slur word
3. categories - which category does this slur word fall under (gender, caster religion etc)
4. appropriated - whether the slur word is used in its direct context or not
5. appropriation_context - if used in the correct context, explain it. 

A user can Add, Edit, Update and Delete a Slur Word they created. 

**Steps to test the PR**
1. Pull this PR
2. Make sure the npm dependencies are installed, if not do so using `npm install`
3. Add `http://localhost:3000 ws://localhost` to the manifest file and make sure you have a `development.env` file setup. 
4. start the backend and frontend as per your OS by following the steps in the wiki. Make sure you backend is up and running. 
5. open the database GUI at `localhost:8080` and see if the tables can be seen. If not do the migrations using `npx sequelize db:migrate` . You should be able to see the two new tables "SLUR" and "CATEGORY". 
6. Everything should now be setup, load the `dist` folder in your desired browser and test the PR. 

**Expected behaviour**
A user can do the CRUD operations for a slur 

**Screenshots**
<img src="https://github.com/tattle-made/Uli/assets/56875084/9ef091e1-9cae-44bb-9a9a-ebf18594ad70" width="350" height="350" alt="Screenshot (503)">
<img src="https://github.com/tattle-made/Uli/assets/56875084/f5f16a4e-35c5-41c3-98a2-0a8e8aa05aad" width="350" height="350" alt="Screenshot (502)">


### **Things to Check closely in a Code Review**
1. all the API requests in the `api-server/index.js` file. I am no longer sending a 501 error in the GET request when there are no slurs found. I saw that when there were no slurs in the table, the GET request was sending an empty array. 
2. ~Check the CRUD API requests when called in frontend, I want to just make sure I am not giving away any user information that could be a potential security bug in the future.~
3. ~The multi-option `categories` is being handled separately; this is causing a `Warning: A component is changing a controlled input to be uncontrolled. This is likely caused by the value changing from a defined to undefined, which should not happen.` I need some help in fixing this warning.~
4. what to do about the new permissions in `manifest.json` files.

### **Potential Things to be added and Worked on**
1. ~Better UI, this will come in shape as final requirements come in~
2. ~Add transactions to PUT and DELETE requests.~ 
3. ~Add a loading bar when CRUD is happening on a Slur~. 
5. ~Modify the delete and edit button to emoji symbol~
6. ~When there are no Slurs for a user, the Add button should be centered at the page below the text.~ 
7. ~test axios requests using Jest~
8. ~Add notifications when slur is created, edited and deleted.~ 
9. ~Add like a warning in the UI if the metadata isn't filled up~
10. Add a go back symbol in the go back text in `Add Slur` form, also make the link a Button. Also add pop-ups, if the data is not saved. 
11. ~sit with denny to fix the true false radio button issuem - THE FORM STATE IS NOT BEING UPDAETD CORRECTELY~
12. ~Ask denny about multiple select dropdown from sematic UI - https://react.semantic-ui.com/modules/dropdown/
https://www.reddit.com/r/reactjs/comments/vtgbai/comparison_of_ui_libraries_for_react/~
~Ask about Tailwind UI - its pretty light - https://tailwind-elements.com/docs/standard/forms/select/~
13. ~Why isn't Grommet Multiple Select work?~
14. ~Add a red star to required questions.~ 
15. ~how do de-select a radio button option - especially for this - If, Appropriated, Is it by Community or Others?~
16. ~`git commit -m "message" --no-verify`~
17. ~button cannot have red background color, why?~
18. ~Add Slur button coming for each slur card because of getSlurs.map() look into this~
19. ~Fix overflow slur issue in card js - fixed it by truncating text~
20. ~ASK denny - the jest testing in npm is being written as npm run test-jest, do I move it to npm run test~
21. ~Should I truncate all the text or should I use `wrap={true}`~
22. ask denny - how to change anchor color in theme
23. Convert all text to hindi and tamil
24. ask denny - better color selection for everything
25. make the CARD using the grommet component.
26. ~make sure slurs appear in descinding order in the slur index page~
27. in enable slur replace, when we cancel the refresh, the notification of saved still pops up, why so?
